### PR TITLE
[Epic] Metric label support

### PR DESCRIPTION
**Pull request by octo**
This PR changes the way metrics are identified. Since this is a fundamental change, it is quite complex and large. I'll try to summarize how the new API works.

Each metric is identified by a *metric name* and a *label set* (a list of key/value pairs, keys must be unique). I'd like to add "resource labels", similar to [the OpenTelemetry "Resource" type](https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/resource/v1/resource.proto), but that's still to do.

## Data structures

There are two central data structures:

*   `metric_family_t` holds the metric type, metric name, and a list of metrics.
*   `metric_t` holds the value, a label set, collection time, interval, and a pointer to the parent *metric family*.

Read plugins generate and dispatch a `metric_family_t` to the daemon, write plugins receive `metric_family_t`. So in many regards, `metric_family_t` takes the role of the previous `value_list_t`.

Other functions take a `metric_t` instead, e.g. `uc_get_rate()` which converts a counter value to a rate, and some formatting / serialization functions.

## Plugin status

*→ [collectd 6 project plan](https://docs.google.com/spreadsheets/d/1ss4NJeJ00CwAmGgIRGQ2EJM9LFifPPrJcq_c0tGN_vY/edit#gid=741037669)*

The following plugins have been updated and work with label metric (on Linux):

*   cpu
*   memory
*   write\_http
*   write\_log
*   write\_stackdriver

There are 173 plugins total.

*   84 compile successfully
*   45 fail to compile
*   44 are unknown (can't build on my development machine)

## Compiling

For the time being, I recommend to compile with:

```sh
./configure --disable-all-plugins --enable-cpu --enable-memory --enable-write_http […]
#           ^^^^^^^^^^^^^^^^^^^^^
```

Issue: #3094

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -439,28 +436,33 @@ libplugin_mock_la_SOURCES = \
 	src/daemon/utils_time.h
 
 libplugin_mock_la_CPPFLAGS = $(AM_CPPFLAGS) -DMOCK_TIME
-libplugin_mock_la_LIBADD = libcommon.la libignorelist.la $(COMMON_LIBS)
+libplugin_mock_la_LIBADD = libcommon.la libignorelist.la libmetadata.la $(COMMON_LIBS)
+#libplugin_mock_la_LIBADD = libcommon.la libignorelist.la $(COMMON_LIBS)
```
Comment (mrunge) [inline Makefile.am:440]: Can we get rid of this line? Looks like an oversight.

Comment (octo) [reply Makefile.am]: Done.

**Review by mrunge — APPROVED** [review]
Thank you!

Let's get this merged and handle any fall-out once we discover issues.

**Comment by octo** [comment]
Thanks Matthias!

**Pull request merged**